src/pipecat/services/aws_nova_sonic/aws.py

```diff
@@ -0,0 +1,973 @@
+    OpenAILLMContext,
+    OpenAILLMContextFrame,
+)
+from pipecat.processors.frame_processor import FrameDirection
+from pipecat.services.aws_nova_sonic.context import (
+    AWSNovaSonicAssistantContextAggregator,
+    AWSNovaSonicContextAggregatorPair,
+    AWSNovaSonicLLMContext,
+    AWSNovaSonicUserContextAggregator,
+    Role,
+)
+from pipecat.services.aws_nova_sonic.frames import AWSNovaSonicFunctionCallResultFrame
+from pipecat.services.llm_service import LLMService
+from pipecat.utils.time import time_now_iso8601
+
+
+class AWSNovaSonicUnhandledFunctionException(Exception):
+    pass
+
+
+class ContentType(Enum):
+    AUDIO = "AUDIO"
+    TEXT = "TEXT"
+    TOOL = "TOOL"
+
+
+class TextStage(Enum):
+    FINAL = "FINAL"  # what has been said
+    SPECULATIVE = "SPECULATIVE"  # what's planned to be said
+
+
+@dataclass
+class CurrentContent:
+    type: ContentType
+    role: Role
+    text_stage: TextStage  # None if not text
+    text_content: str  # starts as None, then fills in if text
+
+    def __str__(self):
+        return (
+            f"CurrentContent(\n"
+            f"  type={self.type.name},\n"
+            f"  role={self.role.name},\n"
+            f"  text_stage={self.text_stage.name if self.text_stage else 'None'}\n"
+            f")"
+        )
+
+
+class AWSNovaSonicLLMService(LLMService):
+    # Override the default adapter to use the AWSNovaSonicLLMAdapter one
+    adapter_class = AWSNovaSonicLLMAdapter
+
+    def __init__(
+        self,
+        *,
+        # TODO: if we have instruction here as an alternative to using context, we should do the same for tools...right?
+        secret_access_key: str,
+        access_key_id: str,
+        region: str,
+        model: str = "amazon.nova-sonic-v1:0",
+        voice_id: str = "matthew",  # matthew, tiffany, amy
+        system_instruction: Optional[str] = None,
+        tools: Optional[ToolsSchema] = None,
+        send_transcription_frames: bool = True,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self._secret_access_key = secret_access_key
+        self._access_key_id = access_key_id
+        self._region = region
+        self._model = model
+        self._client: BedrockRuntimeClient = None
+        self._voice_id = voice_id
+        self._system_instruction = system_instruction
+        self._tools = tools
+        self._send_transcription_frames = send_transcription_frames
+        self._context: AWSNovaSonicLLMContext = None
+        self._stream: DuplexEventStream[
+            InvokeModelWithBidirectionalStreamInput,
+            InvokeModelWithBidirectionalStreamOutput,
+            InvokeModelWithBidirectionalStreamOperationOutput,
+        ] = None
+        self._receive_task = None
+        self._prompt_name = None
+        self._input_audio_content_name = None
+        self._content_being_received = None
+        self._assistant_is_responding = False
+        self._context_available = False
```

Comment:
If `self._context` is `Optional`, do we still need this ?

Reply:
Fixed. A lot cleaner, thanks for the suggestion!